cocos/core/renderer/core/memory-pools.ts

```diff
@@ -0,0 +1,236 @@
+  used for developing games. You are not granted to publish, distribute,
+  sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+/**
+ * @hidden
+ */
+
+import { DEBUG } from 'internal:constants';
+import { GFXRasterizerState, GFXDepthStencilState, GFXBlendState, IGFXBindingLayoutInfo, GFXDevice, GFXBindingLayout, IGFXShaderInfo, GFXShader } from '../../gfx';
+import { NativeBufferPool, NativeObjectPool } from './native-pools';
+
+interface TypedArrayConstructor<T> {
+    new(buffer: ArrayBufferLike, byteOffset: number, length?: number): T;
+    readonly BYTES_PER_ELEMENT: number;
+}
+
+export class BufferPool<T extends TypedArray> {
+
+    // naming convension:
+    // this._bufferViews[chunk][entry][element]
+
+    private _viewCtor: TypedArrayConstructor<T>;
+    private _elementCount: number;
+    private _entryBits: number;
+
+    private _stride: number;
+    private _entriesPerChunk: number;
+    private _entryMask: number;
+    private _chunkMask: number;
+    private _poolFlag: number;
+
+    private _arrayBuffers: ArrayBuffer[] = [];
+    private _freelists: number[][] = [];
+    private _bufferViews: T[][] = [];
+
+    private _nativePool: NativeBufferPool;
+
+    constructor (dataType: PoolDataType, viewCtor: TypedArrayConstructor<T>, elementCount: number, entryBits = 8) {
+        this._viewCtor = viewCtor;
+        this._elementCount = elementCount;
+        this._entryBits = entryBits;
+        this._nativePool = new NativeBufferPool(dataType);
+
+        const bytesPerElement = viewCtor.BYTES_PER_ELEMENT || 1;
+        this._stride = bytesPerElement * elementCount;
+        this._entriesPerChunk = 1 << entryBits;
+        this._entryMask = this._entriesPerChunk - 1;
+        this._poolFlag = 1 << 30;
+        this._chunkMask = 0xffffffff & ~(this._entryMask | this._poolFlag);
+    }
+
+    public alloc () {
+        let i = 0;
+        for (; i < this._freelists.length; i++) {
+            const list = this._freelists[i];
+            if (list.length) {
+                const j = list[list.length - 1]; list.length--;
+                return (i << this._entryBits) + j + this._poolFlag;
+            }
+        }
+        // add a new chunk
+        const buffer = this._nativePool.getChunkArrayBuffer(this._stride * this._entriesPerChunk);
```

Comment:
see  https://github.com/cocos-creator/cocos2d-x-lite/pull/2658

Reply:
Not sure about the necessity of `allocateNewChunk`. What's the benefit of ensuring the same chunk size?

Reply:
> What's the benefit of ensuring the same chunk size? Besides it won't be different anyways since the js pool is responsible for that

If JS assign chunk size for each chunk, I need another array for size of chunks, and to ensure all chunks satisfy entry bits restriction, I need to do more verifications. I don't see any needs to have different chunk size for the same type of pool.

Reply:
And getChunkArrayBuffer is only for retrieve the existing chunk with chunk id, not for creation.